crates/apollo-compiler/src/validation/value.rs

```diff
@@ -0,0 +1,263 @@
+            // When expected as an input type, only integer input values are
+            // accepted. All other input values, including strings with numeric
+            // content, must raise a request error indicating an incorrect
+            // type. If the integer input value represents a value less than
+            // -231 or greater than or equal to 231, a request error should be
+            // raised.
+            // When expected as an input type, any string (such as "4") or
+            // integer (such as 4 or -4) input value should be coerced to ID
+            Value::Int { value: int, .. } => match &type_def {
+                TypeDefinition::ScalarTypeDefinition(scalar) => {
+                    if scalar.is_int() || scalar.is_float() || scalar.is_id() {
+                        if int.to_i32_checked().is_none() {
+                            diagnostics.push(
+                                ApolloDiagnostic::new(
+                                    db,
+                                    val.loc().into(),
+                                    DiagnosticData::IntCoercionError {
+                                        value: int.get().to_string(),
+                                    },
+                                )
+                                .label(Label::new(
+                                    val.loc(),
+                                    "cannot be coerced to an 32-bit integer",
+                                )),
+                            )
+                        }
+                    } else {
+                        diagnostics.push(unsupported_type!(db, val, ty));
+                    }
+                }
+                _ => diagnostics.push(unsupported_type!(db, val, ty)),
+            },
+            // When expected as an input type, both integer and float input
+            // values are accepted. All other input values, including strings
+            // with numeric content, must raise a request error indicating an
+            // incorrect type.
+            Value::Float { .. } => match &type_def {
+                TypeDefinition::ScalarTypeDefinition(scalar) => {
+                    if !scalar.is_float() {
+                        diagnostics.push(unsupported_type!(db, val, ty));
+                    }
+                }
+                _ => diagnostics.push(unsupported_type!(db, val, ty)),
+            },
+            // When expected as an input type, only valid Unicode string input
+            // values are accepted. All other input values must raise a request
+            // error indicating an incorrect type.
+            // When expected as an input type, any string (such as "4") or
+            // integer (such as 4 or -4) input value should be coerced to ID
+            Value::String { .. } => match &type_def {
+                TypeDefinition::ScalarTypeDefinition(scalar) => {
+                    if scalar.is_int() || scalar.is_float() || scalar.is_boolean() {
```

Comment:
the other branches all check if the assignment is _valid_ and otherwise report an error, but this checks if the assignment is _invalid_. it's a bit surprising and i think its more obvious to check all the ways that strings *can* be used instead of the ways they cannot

Reply:
oh yea, definitely, let me change this

Reply:
ohhhh i remember why - this is like this because of custom scalars. custom scalars can be coerced into a string, unless otherwise specified. 

so when you have a directive like this, it passes this validation.
```graphql
directive @join__field(graph: join__Graph, requires: join__FieldSet, provides: join__FieldSet) on FIELD_DEFINITION

scalar join__FieldSet

enum join__Graph {
  REVIEWS
}

type Query {
  retailPrice: String @join__field(graph: REVIEWS, requires: "price")
}
```

Reply:
specifically looking at `requires: "price"`